### PR TITLE
[MXNET-1438] Adding SDML loss function

### PR DESCRIPTION
**Pull request by anjishnu**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- [x] Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [x] Code is well-documented: 
- [x] For user-facing API changes, API doc string has been updated. 
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html

### Changes ###
- [x] Added new loss for information retrieval - Smoothed Deep Metric Learning Loss with relevant API doc.

## Comments ##
- Added loss function outlined in https://arxiv.org/abs/1905.12786 - example to follow.

### CODE REVIEW

**Comment by haojin2** [comment]
@anjishnu Please address the sanity errors: http://jenkins.mxnet-ci.amazon-ml.com/job/mxnet-validation/job/sanity/job/PR-17298/1/display/redirect . Also can you randomize your unit test to ensure that we're covering more numerically different cases?

**Comment by anjishnu** [comment]
Sure will address the sanity cases - 

Can you give an example of a unit test that is appropriately randomized so I can base it on that?

**Comment by haojin2** [comment]
@anjishnu Like this: https://github.com/apache/incubator-mxnet/pull/17298/files#diff-b129d416cc6980021cb35420a66e4d3aR334

**Comment by anjishnu** [comment]
It looks a little tricky to port this into the 'fit' and 'score' paradigm since this is a retrieval specific loss function which uses the other elements in a batch as implicit negative samples - and I'm not sure how cleanly it fits into the Module API for this kind of test. 

The loss only guarantees that associated pairs will be closer in distance space after learning then non-associated pairs. Maybe I can write something equivalent using the gluon API, to train a small network and ensure it learns the right associations. I'll come up with a proposal shortly.

**Comment by haojin2** [comment]
@anjishnu I was actually only asking for getting the input data randomized with the original version of test code untouched.

**Comment by anjishnu** [comment]
@haojin2 if I randomized the input data in the original test code the losses would would have different values during each run (SDML loss imposes a distribution over the relative distances of data points in a minibatch) - so you cannot compare the output against precomputed loss values any more - thus the original unit test procedure cannot be reused.

That's why I added a test that fits a toy model to some toy data instead. The current test was running in ~50 ms on my machine on CPU. Would love to hear your thoughts on how to improve on this.

**Comment by anjishnu** [comment]
Verifying that mention of this PR in #17292 is a typo - author meant to refer to #17208 - we are not breaking horovod

**Comment by anjishnu** [comment]
@haojin2 Are there any serious concerns with the new unit test?

**Comment by haojin2** [comment]
@anjishnu I'm only concerned about if there's any flakiness in such a test, to verify, please try
`MXNET_TEST_COUNT=500 nosetests tests/python/unittest/test_loss.py:test_sdml_loss` on your end and lemme know if it passes.

**Comment by anjishnu** [comment]
@haojin2 
With the original confirguration I was getting around 1 in 100 or 1 in 200 failures, which is not acceptable for a unit test.

Changing the hyperparameters (e.g. increasing dimensionality, lowering N) does make it more robust (tried up to MXNET_TEST_COUNT=2000 runs) but there's always a chance of failure

Other options:
- Go with the original precomputed values approach, but maybe over a larger range of precomputed values.

- Set up a similar test ... but use correlated data instead of using purely random data:
   - Generate the positive samples that are actually correlated to the paired data rather than purely random.
   - Always use some set of samples that are known to converge successfully
- Add a test to see loss reduction (e.g. to <0.05 as in many of the other tests)

Which of these would be your preferred approach - I have tested them all to MXNET_TEST_COUNT=2000

**Comment by anjishnu** [comment]
@haojin2 could you take a look at the latest version - the new unit test (which looks at loss reduction) is passing MXNET_TEST_COUNT=2000

**Comment by haojin2** [comment]
@anjishnu it seems ok. I've re-triggered your failed website build, I think it's good for merge after that passes.

**Comment by anjishnu** [comment]
@haojin2 ok, thanks! sounds good!

**Comment by haojin2** [comment]
@anjishnu Merged, thx for the contribution.

File: python/mxnet/gluon/loss.py [hunk 1]
```diff
@@ -930,3 +930,118 @@ def _cosine_similarity(self, F, x, y, axis=-1):
...
+        Given the question president of France in batch 1 the model will
+        learn to predict french president comparing it with all the other
+        vectors in batch 2
+        """
+        batch_size = x1.shape[0]
+        labels = self._compute_labels(F, batch_size)
+        distances = self._compute_distances(x1, x2)
+        log_probabilities = F.log_softmax(-distances, axis=1)
+        # multiply for the number of labels to obtain the correct loss (gluon kl_loss averages instead of sum)
+        return self.kl_loss(log_probabilities, labels.as_in_context(distances.context)) * batch_size
```
Comment (szha) [inline python/mxnet/gluon/loss.py:1043]: this doesn't work in sym. 1.x will need to be fixed while 2.0 will be switched to deferred compute mode
